src/ApnP8.php

```diff
@@ -0,0 +1,337 @@
+                    $running++;
+                }
+            }
+        }
+
+        if ($this->autoCloseConnections) {
+            curl_multi_close($mh);
+            $this->curlMultiHandle = null;
+        }
+
+        //Set the global feedback
+        $this->setFeedback(json_decode(json_encode($responseCollection)));
+
+        return $responseCollection;
+    }
+
+    /**
+     * Get Url for APNs production server.
+     *
+     * @param Notification $notification
+     * @return string
+     */
+    private function getProductionUrl(string $deviceToken)
+    {
+        return self::APNS_PRODUCTION_SERVER . $this->getUrlPath($deviceToken);
+    }
+
+    /**
+     * Get Url for APNs sandbox server.
+     *
+     * @param Notification $notification
+     * @return string
+     */
+    private function getSandboxUrl(string $deviceToken)
+    {
+        return self::APNS_DEVELOPMENT_SERVER . $this->getUrlPath($deviceToken);
+    }
+
+    /**
+     * Get Url path.
+     *
+     * @param Notification $notification
+     * @return mixed
+     */
+    private function getUrlPath(string $deviceToken)
+    {
+        return str_replace("{token}", $deviceToken, self::APNS_PATH_SCHEMA);
+    }
+
+    /**
+     * Decorate headers
+     *
+     * @return array
+     */
+    public function decorateHeaders(array $headers): array
+    {
+        $decoratedHeaders = [];
+        foreach ($headers as $name => $value) {
+            $decoratedHeaders[] = $name . ': ' . $value;
+        }
+        return $decoratedHeaders;
+    }
+
+    /**
+     * @param $token
+     * @param array $message
+     * @param $request
+     * @param array $deviceTokens
+     */
+    public function prepareHandle($deviceToken, array $message)
+    {
+        $uri = false === $this->config['dry_run'] ? $this->getProductionUrl($deviceToken) : $this->getSandboxUrl($deviceToken);
+        $headers = $message['headers'];
+        $headers['authorization'] = "bearer {$this->generateJwt()}";
+        unset($message['headers']);
+        $body = json_encode($message);
+
+        $options = [
+            CURLOPT_HTTP_VERSION => CURL_HTTP_VERSION_2,
+            CURLOPT_URL => $uri,
+            CURLOPT_PORT => self::APNS_PORT,
+            CURLOPT_POST => true,
+            CURLOPT_POSTFIELDS => $body,
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_TIMEOUT => 10,
+            CURLOPT_HEADER => true,
+            CURLOPT_SSL_VERIFYPEER => true
+        ];
+
+        $ch = curl_init();
+
+        curl_setopt_array($ch, $options);
+        if (!empty($headers)) {
+            curl_setopt($ch, CURLOPT_HTTPHEADER, $this->decorateHeaders($headers));
+        }
+
+        // store device token to identify response
+        curl_setopt($ch, CURLOPT_PRIVATE, $deviceToken);
+
+        return $ch;
+    }
+    
+    protected function generateJwt() {
+        $key = openssl_pkey_get_private('file://'.$this->config['key']);
```

Comment:
I would suggest to not access the file system here to avoid any complications at runtime for the following reasons: 
- Directory structure can look different in other projects
- Certificates should be placed in Laravel's `/storage` folder and read from there using `storage_path()` helper

Suggestion: Could you move the `openssl_pkey_get_private` method into the `config.php` file and access it from the storage path?

Reply:
I was just following the folder structure that exists nowadays. 

There is the iosCertificates folder right there, I didn't place the .p8 file into the folder because it wasn't technically a certificate, but there would be no problem placing it "by default" inside a new folder. This route can be set in the config anyway, in case another project has a different structure.